hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/HoodieTimelineArchiver.java

```diff
@@ -452,107 +431,137 @@ private Stream<HoodieInstant> getCommitInstantsToArchive() throws IOException {
       // without the replaced files metadata on the timeline, the fs view would expose duplicates for readers.
       // Meanwhile, when inline or async clustering is enabled, we need to ensure that there is a commit in the active timeline
       // to check whether the file slice generated in pending clustering after archive isn't committed.
       Option<HoodieInstant> oldestInstantToRetainForClustering =
           ClusteringUtils.getOldestInstantToRetainForClustering(table.getActiveTimeline(), table.getMetaClient());
+      oldestInstantToRetainCandidates.add(oldestInstantToRetainForClustering);
+
+      // 4. If metadata table is enabled, do not archive instants which are more recent than the last compaction on the
+      // metadata table.
+      if (table.getMetaClient().getTableConfig().isMetadataTableAvailable()) {
+        try (HoodieTableMetadata tableMetadata = HoodieTableMetadata.create(table.getContext(), config.getMetadataConfig(), config.getBasePath())) {
+          Option<String> latestCompactionTime = tableMetadata.getLatestCompactionTime();
+          if (!latestCompactionTime.isPresent()) {
+            LOG.info("Not archiving as there is no compaction yet on the metadata table");
+            return Collections.emptyList();
+          } else {
+            LOG.info("Limiting archiving of instants to latest compaction on metadata table at " + latestCompactionTime.get());
+            oldestInstantToRetainCandidates.add(Option.of(new HoodieInstant(
+                HoodieInstant.State.COMPLETED, COMPACTION_ACTION, latestCompactionTime.get())));
+          }
+        } catch (Exception e) {
+          throw new HoodieException("Error limiting instant archival based on metadata table", e);
+        }
+      }
+
+      // 5. If this is a metadata table, do not archive the commits that live in data set
+      // active timeline. This is required by metadata table,
+      // see HoodieTableMetadataUtil#processRollbackMetadata for details.
+      if (table.isMetadataTable()) {
+        HoodieTableMetaClient dataMetaClient = HoodieTableMetaClient.builder()
+            .setBasePath(HoodieTableMetadata.getDatasetBasePath(config.getBasePath()))
+            .setConf(metaClient.getHadoopConf())
+            .build();
+        Option<HoodieInstant> qualifiedEarliestInstant =
+            TimelineUtils.getEarliestInstantForMetadataArchival(
+                dataMetaClient.getActiveTimeline(), config.shouldArchiveBeyondSavepoint());
+
+        // Do not archive the instants after the earliest commit (COMMIT, DELTA_COMMIT, and
+        // REPLACE_COMMIT only, considering non-savepoint commit only if enabling archive
+        // beyond savepoint) and the earliest inflight instant (all actions).
+        // This is required by metadata table, see HoodieTableMetadataUtil#processRollbackMetadata
+        // for details.
+        // Todo: Remove #7580
```

Comment:
After this PR, #7580 is no useful, consider remove or simplify it

Reply:
The incremental cleaning needs to deser the clean plan, what if the cleaning has been archived, is it affected?

Reply:
> The incremental cleaning needs to deser the clean plan, what if the cleaning has been archived, is it affected?

see the instant order:  commit1, commit2, commit3, clean1, commit4 ...

We assume clean1 is the last clean (using for incremental cleaning), commit1 is the earliest commit to retain in the clean plan, incremental cleaning is only valid when commit1 exists, and only when commit1 is archived, clean1 can be archived under this PR, so incremental cleaning is always valid as long as commit1 is not archived.

As for whether commit1 is archived, we can create an another jira to discuss whether it will happen, is also simple to realize it, just add another judgment in `getCommitInstantsToArchive `. Of course, we currently have a fallback mechanism for the worst case (#8373)

Reply:
>  incremental cleaning is only valid when commit1 exists, and only when commit1 is archived

Can you elaborate this a little?

Reply:
![image](https://github.com/apache/hudi/assets/37108074/f115d1d7-8241-4922-ab01-c6f344e93d32)

assume cleanMetadata.getEarliestCommitToRetain() is commit1, and the ts of commt1 < the ts of clean1. Under this PR, and only when commit1 is archived, clean1 can be archived

Reply:
We should keep at least clean commit on the active timeline right? What about the rollback.